experiment/clustersecretbackup/main.go

```diff
@@ -0,0 +1,237 @@
+	fs.BoolVar(&o.update, "update", false, "Controls whether update existing secret or not, if false then secret will only be created")
+	fs.BoolVar(&o.dryRun, "dryrun", false, "Controls whether this is dry run or not")
+	fs.Parse(args)
+
+	return o
+}
+
+// validateFlags validates the command-line flags.
+func (o *options) validateFlags() error {
+	if len(o.project) == 0 {
+		return errors.New("--project must be provided")
+	}
+	if len(o.clusterContext) == 0 {
+		return errors.New("--cluster-context must be provided")
+	}
+	return nil
+}
+
+func newClient(o options) (*client, error) {
+	secretmanagerClient, err := secretmanager.NewClient(o.project)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating secret manager client: %w", err)
+	}
+
+	kubeClient, err := newKubeClients(o.clusterContext)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating kube client: %w", err)
+	}
+	return &client{
+		kubeClient:          kubeClient,
+		secretmanagerClient: secretmanagerClient,
+		options:             o,
+	}, nil
+}
+
+func newKubeClients(clusterContext string) (ctrlruntimeclient.Client, error) {
+	var loader clientcmd.ClientConfigLoader
+	loader = clientcmd.NewDefaultClientConfigLoadingRules()
+
+	cfg, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		loader, &clientcmd.ConfigOverrides{
+			// Enforcing clusterContext in the full form of cluster context,
+			// instead of short names for kubectl.
+			Context:        api.Context{Cluster: clusterContext},
+			CurrentContext: clusterContext,
+		}).ClientConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed create rest config: %v ------ Did you supply the full form of cluster context name instead of short hand?", err)
+	}
+	return ctrlruntimeclient.New(cfg, ctrlruntimeclient.Options{})
+}
+
+// process merges secret into a new secret for write.
+func (c *client) updateSingleSecret(ctx context.Context, clusterSecret *corev1.Secret) error {
+	secretID := c.gsmSecretName(clusterSecret)
+	// Google secret manager expects pure string instead of map[string][]byte,
+	// so translate *corev1.Secret.Data to map[string]string, this will also be
+	// what kubernetes external secret expects.
+	stringData := map[string]string{}
+	for key, val := range clusterSecret.Data {
+		stringData[key] = string(val)
+	}
+	payload, err := json.Marshal(stringData)
+	if err != nil {
+		return fmt.Errorf("failed marshal secret %s: %w", clusterSecret.Name, err)
+	}
+	log := logrus.WithFields(logrus.Fields{
+		"project":     c.project,
+		"cluster":     c.clusterContext,
+		"namespace":   clusterSecret.Namespace,
+		"secret-name": clusterSecret.Name,
+		"gsm-secret":  secretID,
+	})
+	if sat := "kubernetes.io/service-account-token"; string(clusterSecret.Type) == sat {
+		log.Infof("Skipping: the secret type is %s", sat)
+		return nil
+	}
+	if c.dryRun {
+		log.Info("[Dryrun]: backing up secret")
```

Comment:
nit: We could still try to list secrets in dry run mode.
Consider handling dry run from the secretmanagerclient level to ensure all mutating calls respect dryrun without the caller worrying about dryrun checks. That is what Prow's GitHub client does.

Reply:
good idea and done